### PR TITLE
Activate C++20 and add clang-tidy checks

### PR DESCRIPTION
**Pull request by nsavoire**
# What does this PR do?

* Compile in C++20 mode
* Enable clang-tidy checks during compilation when`-DENABLE_CLANG_TIDY=ON` is passed to cmake
* Fix some clang-tidy warnings

# Motivation

* Add more static analysis tools
* Benefit from latest C++ version

# Additional Notes

This is the section to put technical guidance/constraints, call out potential regressions, cite sources, and in general offer some exposition on the _how_ of the PR.  This section doesn't need to be incredibly detailed, but it makes life easier for reviewers!

# How to test the change?

Describe here in detail how the change can be validated.  This is a great section to call out specific tests you've added or improved, or to acknowledge code sections which are particularly difficult to test.

### CODE REVIEW

File: cmake/ExtendBuildTypes.cmake [hunk 1]
```diff
@@ -6,6 +6,9 @@ if(NOT CMAKE_BUILD_TYPE)
   set(CMAKE_BUILD_TYPE "Release" CACHE STRING "CMake build type" FORCE)
 endif()
 
+set(CMAKE_CXX_STANDARD 20)
```
Comment (r1viollet) [inline cmake/ExtendBuildTypes.cmake:9]: Global C++ version is good. I had trouble finding this 👍

Comment (nsavoire) [reply cmake/ExtendBuildTypes.cmake:9]: I don't know if this is the right place to put it, if you have a better idea, I am all ears :)

Comment (r1viollet) [reply cmake/ExtendBuildTypes.cmake:9]: I was thinking at the start of the general cmake file, but this is also fine.

**Review by r1viollet — APPROVED** [review]
LGTM
Yay string views are here !

**Pull request merged**